### ISSUE TITLE
driver causes I2C issues with Unexpected Maker RTC shield (MCP7940)

### ISSUE THREAD

**Issue by PaulskPt** [peter-l5/SH1107#6]
Hi Peter, I like your SH1107 driver because of the various additions you made. Unfortunately I encountered serious problems with my current setup: board: Unexpected Maker, ESP32-S3 Feather S3 (P4), display:: Adafruit 1.12in 128x128px OLED mono (SH1107), RTC shield: Unexpected Maker using an MCP7940. The FeatherS3 is flashed with: "MicroPython v1.21.0 on 2023-10-06". When I use your driver with this setup the MCP1107 the _get_time() and some other functions like _set_bit() report timeout or device not found (ENODEV) errors. When I use a more basic SH1107 driver by @nemart69 [sh1107](https://github.com/nemart69/sh1107-micropython), then the MCP7940 works OK, however the nemart69 driver lacks options like rotation. I tried to figure out what is the reason of your driver is causing problems for a 2nd device on the same I2C bus. I did not find the source of the problem.  I just want to inform you about my experiences.

I created a shared folder on my dropbox.  There are all the files I use for my project and also the two screenshots and two REPL output files: [shared_folder](https://www.dropbox.com/scl/fo/nv6rmsw5jl6w4q2msmet3/h?rlkey=6fpgsupe2xyrw20s92u701m6l&dl=0)

**Comment by peter-l5** [comment]
hi @PaulskPt,
thanks for raising this. I've not come across unexpected maker boards before (they look interesting) and definitely don't have those that you mention so won't be able to reproduce the error.
Thank you for sharing the files. I have had a look at them but couldn't see the lines of code that were producing the error. I did wonder if the issue might be a name conflict around the term/object `time`. However I'm not at all certain that this is so. Do you think this could be it?
I have made a branch which comments out references to the time module. I think this should still work except if perhaps you are using a reset pin (which isn't necessary for I2C connections).
Maybe you could try that?
[issue 6 branch](https://github.com/peter-l5/SH1107/blob/issue6_idea/sh1107.py)
alternatively you might try renaming the MCP7940 class' time() property?
let me know if this helps at all?

**Comment by PaulskPt** [comment]
Hi Peter, thank you for your rapid reply. I dont think it is a naming problem. I use the MCP4970.py successfully in Circuitpython (be it that the i2c functions are different and using I2C bus lock - unlock commands. 
I go to see your branch.  I am not using a reset pin. I keep you posted.

**Comment by PaulskPt** [comment]
In the file MCP7940.py class MCP7940 I have renamed 'time' (getter & setter) renamed to mcptime.
Function _get_time() I have renamed to _mcpget_time(). I used it with your sh1107.py branch issue6 driver. No difference.
in the sh1107_i2c class, functions write_command() and write_data() I added some code to catch the OSErrors and print the command resp. the data sent.. See file: sh1107_issue6.py

**Comment by nemart69** [comment]
> Hi Peter, I like your SH1107 driver because of the various additions you made. Unfortunately I encountered serious problems with my current setup: board: Unexpected Maker, ESP32-S3 Feather S3 (P4), display:: Adafruit 1.12in 128x128px OLED mono (SH1107), RTC shield: Unexpected Maker using an MCP7940. The FeatherS3 is flashed with: "MicroPython v1.21.0 on 2023-10-06". When I use your driver with this setup the MCP1107 the _get_time() and some other functions like _set_bit() report timeout or device not found (ENODEV) errors. When I use a more basic SH1107 driver by @nemart69 [sh1107](https://github.com/nemart69/sh1107-micropython), then the MCP7940 works OK, however the nemart69 driver lacks options like rotation. I tried to figure out what is the reason of your driver is causing problems for a 2nd device on the same I2C bus. I did not find the source of the problem. I just want to inform you about my experiences.
> 
> I created a shared folder on my dropbox. There are all the files I use for my project and also the two screenshots and two REPL output files: [shared_folder](https://www.dropbox.com/scl/fo/nv6rmsw5jl6w4q2msmet3/h?rlkey=6fpgsupe2xyrw20s92u701m6l&dl=0)

Hi Paulus, I cannot help with your issue, just a minor comment regarding my driver and rotation - there is no explicit _rotate_ option, however you can initialize a rectangular display both in horizontal (width = 128, height = 64) and vertical (width = 64, height = 128) mode, i.e. you can get it rotated by 90 degrees this way.

**Comment by PaulskPt** [comment]
@nemart69 Thank you for your reaction and your tip about rotation for your driver.

**Comment by peter-l5** [comment]
Thanks for the messages. Having given this some further thought I'm wondering if amending the initialisation parameters might resolve this. The display has about three parameters that affect things like the internal charge/ discharge cycle that lights the pixels. 
A difference with this driver and @nemart69's is in the parameter `_SET_DC_DC_CONVERTER_SF  = const(0xad8d)`. It might help to set this to the power on reset value (see SH1107 data sheet) of `0xad81`. In the code for this display the other values are not set so will take their power on reset values. If changing this one doesn't help then amending the others might. 
I will test this over the coming days, but if you try it me know how it goes. 
Incidentally, I have found that the ENODEV error essentially means Exists NO DEVice. (I'll try and post a reference separately.) I'm wondering if the above setting might cause two high a current draw on the I2C line, thus impacting the RTC, but that's speculation really as I can't claim much expertise at all in that!

**Comment by PaulskPt** [comment]
Thank you very much Peter. I made the change for `_SET_DC_DC_CONVERTER_SF  = const(0xad8d)` 

Below the REPL output: 

```
>>> %Run -c $EDITOR_CONTENT
create mcp object try nr: 1
mcp object created
framebuf is  standard
SH1107.reset() has been called
We are using the SH1107 display driver from peter-I5
We are not using the SH1107_2 display driver from nemart69
i2c devices present:
device nr 1, addres dec:  61, addres hex: 0x3d, OLED display (SH1107)
device nr 2, addres dec: 111, addres hex: 0x6f, RTC shield (MCP7940)

setup():       board unique_id: 84:f7:03:c0:65:40
setup():       MCP7940 not started yet...
setup():       MCP7940 now started
setup():       MCP7940 power failure occurred? False
setup():       MCP7940 backup battery is not enabled
setup():       check: is_12hr: 1
do_connect():  connecting to network...
do_connect():  WiFi connected to 'xxxxx'
MYNTPTIME.ntp_time(): using host: '2.pt.pool.ntp.org'
MYNTPTIME.ntp_time(): msg rcvd from host ( showing just sliced part: list(msg[40:44]) ): [232, 219, 135, 247]
set_time():    Succeeded to update the builtin RTC from an NTP server
set_time():    setting MCP7940 timekeeping regs to: (2023, 10, 19, 11, 48, 23, 3, 292)
MCP7940.mcptime() setter: 23/10/19 11:48:23 (day=3)
MCP7940.mcptime() setter: time check: (2023, 10, 19, 11, 48, 23, 3)
setup():       Microcontroller (utime.localtime()) year = 2023
setup():       MCP7940_RTC datetime year                = 2023
setup():       utime.localtime() result: (2023, 10, 19, 11, 48, 10, 3, 292)
setup():       MCP7940 RTC set to: (2023, 10, 19, 11, 48, 23, 3, 292)

main():        Test: saving and reading datetime data to/from MCP7940 SRAM
MCP7940._mcpget_time(): Error: [Errno 19] ENODEV                                                        <<<==== ENODEV error
main():        Current MCP7940 RTC datetime: Tuesday 2001-01-01 0:00:00 AM. Day of year:   1
```

Other observations/knowledge:
a) about the FeatherS3. The FeatherS3 I am using is version 'P4'. It has 2 STEMMA I2C connectors. One is named (by UM, the manufacturer) 'STEMMA_VERTICAL'. The oither is named 'STEMMA_HORIZONTAL', Both STEMMA I2C connectors use the same pins: 8 and 9. UM is now selling FeatherS3 version P7 (and P8). From version P7 the STEMMA_VERTICAL is connected to Pins 15 and 16. I think that this has also been done for reasons of possible too high current draw. And for being able to define two different I2C buses.
b) When I connect the probes of my Oscilloscope to the I2C pins of the MCP7940 (pins 21 and 22), which are connected to the STEMMA VERTICAL (pins 8 and 9) of the FeatherS3, then the script is reporting a stream of I2C errors. Strange to me because the probes have an impedance of 1 MOhm. I never had such an experience before while doing measurements on microcontrollers with this oscilloscope. So this also gives me the impression that the I2C signal levels are "weak" so to speak. Having no "reserve" to add an oscilloscope probe.

**Comment by PaulskPt** [comment]
Changing the _SET_DC_DC_CONVERTER_SF didn't help.

**Comment by peter-l5** [comment]
Hi @PaulskPt ,
Thank you for trying out that idea and the feedback. 
I have been looking into this further and have created a further branch
[SH1107 v319](https://github.com/peter-l5/SH1107/tree/v319) - perhaps you might like to try this out? 

Info on this and some thoughts from the research are as follows: 
OLED displays can draw quite a substantial current - 100mA or more; 
the display driver chip has an internal charge pump; 
I have experienced EIO errors sometimes with SH1107 - normally when writing to the display immediately after initialisation/ running the constructor. I therefore suspect that there may be a heavy current draw as the display initialises.
The driver had a 100ms delay in the power on sequence - this is suggested on the SH1107 datasheet. Extending it to 200ms seemed to stop these errors completely so I have amended the driver to have a 200ms power on (and awake from sleep delay) by default. I've also parameterised it, so that anyone who doesn't want the delay can initialise it without a delay.  
I tried the SH1107 parameter values used in the @nemart69 driver. However in the displays I tested this with it seemed to result in the black pixels no longer being fully black, so I decided not to adopt that approach. 
I've also tried the SH1107 datasheet power on reset values for clock and charge pump type settings and these seem to work well. So I have amended the setup to these values: overall this results in changing just one parameter (_SET_DC_DC_CONVERTER_SF).
I wondered if the additional delay might sort out the problem that you have experienced? If you are able to try this, I'd be very interested to hear of the outcome.
Incidentally, here is an interesting blog I found on OLED current use: [how much current do oled displays use?](https://bitbanksoftware.blogspot.com/2019/06/how-much-current-do-oled-displays-use.html)

**Comment by peter-l5** [comment]
Probably best ignore the discussion on the _SET_DC_DC_CONVERTER_SF parameter, that's a bit of a red herring here (as already tried as a fix), but included for completeness. 
The startup power delay change though might help?

**Comment by PaulskPt** [comment]
Problem solved Peter. I discovered that when I use the STEMMA_VERTICAL of this FeatherS3 (P4), connecting one of the two I2C devices: the SH1107, then there come a lot of I2C errors. When I connect both of the devices: the SH1107 display and the MCP7940 RTC to a Sparkfun qwiic multi port and one I2C wire from that Sparkfun to the STEMMA_HORIZONTAL connector of the FeatherS3 and also my oscilloscope probes to pins 21 and 22 of the MCP7940, then all runs OK with your sh1107_issue6.py version. I go to check if it also runs OK with your standard sh1107.py driver. Conclusion. I think there is a power problem with this FeatherS3 STEMMA_VERTICAL connector. I go to communicate this with Unexpected Maker,  the developer/manufacturer of the FeatherS3. I put some images of oscillograms and an image of the connection of the two I2C devices through the Sparkfun qwiic multi port (cross). Thank you for your assistance!

Heres is a link to the folder with images of the hardware setup and two REPL output files: [solved](https://www.dropbox.com/scl/fo/p4on15mpq7fqfr2nltet1/h?rlkey=sz7h0ctm79v0nturheonsecmh&dl=0)

P.S.: I checked the hardware setup (through STEMMA_HORIZONTAL with your standard sh1107.py driver. Works OK.

**Comment by PaulskPt** [comment]
Peter, I have to send you my apologies for having created this issue and taking your time. Fault is all mine. I just discovered that in my script I failed to load (import) a library file for the FeatherS3, called: feathers3.py. Together with this I failed to add, in the start of my script, anyway before creating the sh1107 object, making this call: 'feathers3.set_ldo2_power(True)'. Since I have added the statements:

```
... import feathers3
... feathers3.set_ldo2_power(True)
```
the script runs OK, while using your sh1107.py driver.
